frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-builder/__tests__/form-switcher-validation.spec.ts

```diff
@@ -0,0 +1,104 @@
+import * as _ from 'lodash';
+import { initialPipelineFormData } from '../const';
+import {
+  getFormData,
+  handleSanitizeToFormError,
+  safeOmit,
+  sanitizeToForm,
+} from '../form-switcher-validation';
+import { updateYAML, yamlDataBasicPassState } from './switch-to-form-validation-utils-data';
+import { formDataBasicPassState } from './validation-utils-data';
+
+describe('sanitizeToForm', () => {
+  it('should return with formdata if the yamlPipeline provided is valid', async () => {
+    const finalFormData = _.merge(
+      {},
+      initialPipelineFormData,
+      getFormData(formDataBasicPassState, yamlDataBasicPassState),
+    );
+    const sanitizedFormData = await sanitizeToForm(formDataBasicPassState, yamlDataBasicPassState);
+    expect(sanitizedFormData).toMatchObject(finalFormData);
+  });
+
+  it('should return a function if the yamlPipeline provided is invalid', async () => {
+    const inValidYaml = updateYAML('spec.tasks', [
+      { notName: 'test', taskRef: { name: 'external-task' } },
+    ]);
+    const sanitizedFormData = await sanitizeToForm(formDataBasicPassState, inValidYaml);
+    expect(typeof sanitizedFormData).toEqual('function');
```

Comment:
In my opinion, testing the result from the callback returned from `sanitizedFormData` would be similar to testing `handleSanitizeToFormError`.
Expected behavior from `sanitizedFormData` is it should either return `formData` or a callback. but we could check if the callback returns `PipelineBuilderFormValues`, since it is defined in its type def.

Reply:
It's not entirely that. If you look at the code there is a fallback. But I suppose as long as you have a test that verifies returning undefined, we can ignore this test.